chore(amplify): domain name validation

badmintoncryer opened the pull request:
### Issue # (if applicable)

None

### Reason for this change

We can configure amplify domain name but there is no validation for that.

### Description of changes

Add validation for an amplify domain name
- length must be lower than 255 characters
- match with [the regular expression](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-amplify-domain.html#cfn-amplify-domain-domainname)
  - /^(((?!-)[A-Za-z0-9-]{0,62}[A-Za-z0-9])\.)+((?!-)[A-Za-z0-9-]{1,62}[A-Za-z0-9])(\.)?$/

### Description of how you validated changes

Add unit test

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

mazyu36 commented on packages/@aws-cdk/aws-amplify-alpha/lib/domain.ts:
```diff
@@ -131,6 +131,13 @@ export class Domain extends Resource {
     this.subDomains = props.subDomains || [];
 
     const domainName = props.domainName || id;
+    if (domainName.length > 255) {
```
I think we might need to check token.

```typescript
Token.isUnresolved(domainName)
```

badmintoncryer replied:
I keep forgetting to check the token no matter how many times I try... Thank you for the reminder!

mazyu36 commented on packages/@aws-cdk/aws-amplify-alpha/lib/domain.ts:
```diff
@@ -131,6 +131,13 @@ export class Domain extends Resource {
 
     const domainName = props.domainName || id;
+    if (domainName.length > 255) {
+      throw new Error(`Domain name must be 255 characters or less, got: ${domainName.length}`);
```
```suggestion
      throw new Error(`Domain name must be 255 characters or less, got: ${domainName.length} characters.`);
```

I think it would be more helpful to include "characters" in the error message.

mazyu36 commented on packages/@aws-cdk/aws-amplify-alpha/test/domain.test.ts:
```diff
@@ -176,6 +176,55 @@ test('map a branch to the domain root', () => {
+      },
+    ],
+    domainName: 'a'.repeat(256),
+  })).toThrow('Domain name must be 255 characters or less, got: 256');
```
```suggestion
  })).toThrow('Domain name must be 255 characters or less, got: 256 characters.');
```

mazyu36 reviewed (commented):
Thank you for the contribution.
some minor comments.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 98a0dc26b605658879c4ca4e70b7e96903f31d32
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bbfd62b7-bcfd-45ec-b7bc-38fca7c2c85f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

badmintoncryer commented:
@mazyu36 Thank you for your really quick review!! I've addressed all of your comments.

Unfortunately, the `integ.app-custom-domain.ts` you previously created has failed. The following error message appears, but do you know where the example.com shown here is configured?

```sh
/codebuild/output/src3849279982/src/github.com/aws/aws-cdk/packages/@aws-cdk/aws-amplify-alpha/lib/domain.js:30
@aws-cdk/aws-amplify-alpha:             throw new Error(`Domain name must be a valid hostname, got: ${domainName}`);
@aws-cdk/aws-amplify-alpha:             ^
@aws-cdk/aws-amplify-alpha: Error: Domain name must be a valid hostname, got: *.example.com
@aws-cdk/aws-amplify-alpha:     at new Domain (/codebuild/output/src3849279982/src/github.com/aws/aws-cdk/packages/@aws-cdk/aws-amplify-alpha/lib/domain.js:30:19)
@aws-cdk/aws-amplify-alpha:     at App.addDomain (/codebuild/output/src3849279982/src/github.com/aws/aws-cdk/packages/@aws-cdk/aws-amplify-alpha/lib/app.js:151:16)
@aws-cdk/aws-amplify-alpha:     at new TestStack (/codebuild/output/src3849279982/src/github.com/aws/aws-cdk/packages/@aws-cdk/aws-amplify-alpha/test/integ.app-custom-domain.js:28:28)
@aws-cdk/aws-amplify-alpha:     at Object.<anonymous> (/codebuild/output/src3849279982/src/github.com/aws/aws-cdk/packages/@aws-cdk/aws-amplify-alpha/test/integ.app-custom-domain.js:54:15)
@aws-cdk/aws-amplify-alpha:     at Module._compile (node:internal/modules/cjs/loader:1364:14)
@aws-cdk/aws-amplify-alpha:     at Module._extensions..js (node:internal/modules/cjs/loader:1422:10)
@aws-cdk/aws-amplify-alpha:     at Module.load (node:internal/modules/cjs/loader:1203:32)
@aws-cdk/aws-amplify-alpha:     at Module._load (node:internal/modules/cjs/loader:1019:12)
@aws-cdk/aws-amplify-alpha:     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:128:12)
@aws-cdk/aws-amplify-alpha:     at node:internal/main/run_main_module:28:49
@aws-cdk/aws-amplify-alpha: Node.js v18.20.4
@aws-cdk/aws-amplify-alpha:   ERROR      integ.app-custom-domain 0.509s
```

mazyu36 commented:
@badmintoncryer 
Ah...
That domain name is actually a placeholder value defined below.
https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/integ-runner/lib/runner/runner-base.ts#L430

For integ test runs, the actual domain is set in an environment variable. 
However, in the synthesized template, no personal domain names are retained, and only placeholder values remain.
It seems that this placeholder value causes a validation error.

I’m currently considering ways to avoid this, but I haven’t found a good solution yet. 
Changing the placeholder value would require rerunning all integration tests that rely on this placeholder. 
I also thought about modifying the placeholder just for this specific test, but even then, this integration test would still need to be rerun...

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ea43c6ae51773b936f8aaf83fd3086249e9811d0
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8a19af53-5e59-4c51-b645-f1fa0894e4fe%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

badmintoncryer commented:
@mazyu36 Thank you for detail information! I understand very well!

If I modify the referenced value to `example.com` as shown below and then rerun the integ test, would that resolve the issue?

```diff
- const domainName = process.env.CDK_INTEG_DOMAIN_NAME ?? process.env.DOMAIN_NAME;
+ const domainName = process.env.CDK_INTEG_HOSTED_ZONE_NAME ?? process.env.DOMAIN_NAME;
```

This is a good opportunity, so I’ll go ahead and try purchasing a domain too. Haha

badmintoncryer commented:
@mazyu36 I've bought my domain and rerun integ test.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3a74ca110161fac6b77e8fd083d75eeeca041e30
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=32fee7c7-4793-42cb-9be3-112b521a8730%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d43d22f31af9692aaea5516d1647df0821ce2189
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f45da57c-56e8-4caa-822f-a4fc4db6cbba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 23efe87c92e16183733150caff753726f4232847
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e2a64ff9-1bc2-482a-bd86-a809f751acaa%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

mazyu36 reviewed (approved):
Thanks! LGTM

paulhcsun commented:
@mergifyio update

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 91efe523189190c11b8b3a0431c4ad873d189ece
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d72e82e1-3ac8-4e7c-bb79-d336213bef67%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
